File: iosApp/iosApp/Modules/QRCodeScan ✅/QRCodeScanWireframeFactory.swift
// Created by web3d4v on 21/06/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3walletcore

// MARK: QRCodeScanWireframeFactory

protocol QRCodeScanWireframeFactory {
    func make(
        _ parent: UIViewController?,
        context: QRCodeScanWireframeContext
    ) -> QRCodeScanWireframe
}

// MARK: DefaultQRCodeScanWireframeFactory

final class DefaultQRCodeScanWireframeFactory {}

extension DefaultQRCodeScanWireframeFactory: QRCodeScanWireframeFactory {

    func make(
        _ parent: UIViewController?,
        context: QRCodeScanWireframeContext
    ) -> QRCodeScanWireframe {
        DefaultQRCodeScanWireframe(
            parent,
            context: context
        )
    }
}

// MARK: Assembler

final class QRCodeScanWireframeFactoryAssembler: AssemblerComponent {
    
    func register(to registry: AssemblerRegistry) {
        registry.register(scope: .instance) { resolver -> QRCodeScanWireframeFactory in
            DefaultQRCodeScanWireframeFactory()
        }
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/compose/qrcodescan/QRCodeScanWireframeFactory.kt
package com.sonsofcrypto.web3wallet.android.modules.compose.qrcodescan

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.common.AssemblerComponent
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistry
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistryScope
import com.sonsofcrypto.web3walletcore.modules.qrCodeScan.QRCodeScanWireframe
import com.sonsofcrypto.web3walletcore.modules.qrCodeScan.QRCodeScanWireframeContext
import smartadapter.internal.extension.name

interface QRCodeScanWireframeFactory {
    fun make(parent: Fragment?, context: QRCodeScanWireframeContext): QRCodeScanWireframe
}

class DefaultQRCodeScanWireframeFactory: QRCodeScanWireframeFactory {

    override fun make(
        parent: Fragment?, context: QRCodeScanWireframeContext
    ): QRCodeScanWireframe = DefaultQRCodeScanWireframe(
        parent?.let { WeakRef(it) },
        context,
    )
}

class QRCodeScanWireframeFactoryAssembler: AssemblerComponent {

    override fun register(to: AssemblerRegistry) {
        to.register(QRCodeScanWireframeFactory::class.name, AssemblerRegistryScope.INSTANCE) {
            DefaultQRCodeScanWireframeFactory()
        }
    }
}
